File: src/com/ptby/dynamicreturntypeplugin/signatureconversion/CustomSignatureProcessor.java
package com.ptby.dynamicreturntypeplugin.signatureconversion;

import com.intellij.openapi.project.Project;
import com.jetbrains.php.PhpIndex;
import com.jetbrains.php.lang.psi.elements.PhpNamedElement;
import com.ptby.dynamicreturntypeplugin.index.ClassConstantAnalyzer;
import com.ptby.dynamicreturntypeplugin.index.FieldReferenceAnalyzer;
import com.ptby.dynamicreturntypeplugin.index.ReturnInitialisedSignatureConverter;
import com.ptby.dynamicreturntypeplugin.index.VariableAnalyser;

import java.util.Collection;
import java.util.Collections;

import static com.intellij.openapi.diagnostic.Logger.getInstance;

public class CustomSignatureProcessor {
    private final com.intellij.openapi.diagnostic.Logger logger = getInstance( "DynamicReturnTypePlugin" );
    private ReturnInitialisedSignatureConverter returnInitialisedSignatureConverter;
    private ClassConstantAnalyzer classConstantAnalyzer;
    private FieldReferenceAnalyzer fieldReferenceAnalyzer;
    private VariableAnalyser variableAnalyser;


    public CustomSignatureProcessor( ReturnInitialisedSignatureConverter returnInitialisedSignatureConverter,
                                     ClassConstantAnalyzer classConstantAnalyzer,
                                     FieldReferenceAnalyzer fieldReferenceAnalyzer,
                                     VariableAnalyser variableAnalyser ) {
        this.returnInitialisedSignatureConverter = returnInitialisedSignatureConverter;
        this.classConstantAnalyzer = classConstantAnalyzer;
        this.fieldReferenceAnalyzer = fieldReferenceAnalyzer;
        this.variableAnalyser = variableAnalyser;
    }


    public Collection<? extends PhpNamedElement> getBySignature( String signature, Project project ) {
        PhpIndex phpIndex = PhpIndex.getInstance( project );
        CustomMethodCallSignature customMethodCallSignature = CustomMethodCallSignature.OBJECT$.createFromString( signature );
        if ( customMethodCallSignature == null ) {
            return tryFunctionCall( null, signature, phpIndex, project );
        }

        SignatureMatcher signatureMatcher = new SignatureMatcher();
        if ( signatureMatcher.verifySignatureIsDeferredGlobalFunctionCall( customMethodCallSignature ) ||
                signatureMatcher.verifySignatureIsFromReturnInitialiasedLocalObject( customMethodCallSignature ) ) {
            customMethodCallSignature = returnInitialisedSignatureConverter.convertSignatureToClassSignature(
                    customMethodCallSignature, project
            );
        }

        if ( signatureMatcher.verifySignatureIsClassConstantFunctionCall( customMethodCallSignature ) ) {
            return phpIndex.getAnyByFQN(
                    classConstantAnalyzer
                            .getClassNameFromConstantLookup(
                                    customMethodCallSignature
                                            .getRawStringSignature(), project
                            )
            );
        } else if ( signatureMatcher.verifySignatureIsFieldCall( customMethodCallSignature ) ) {
            return fieldReferenceAnalyzer.getClassNameFromFieldLookup( customMethodCallSignature, project );
        } else if ( signatureMatcher.verifySignatureIsMethodCall( customMethodCallSignature ) ) {
            return variableAnalyser.getClassNameFromVariableLookup( customMethodCallSignature, project );
        }

        return tryToDeferToDefaultType( customMethodCallSignature, signature, phpIndex );
    }


    private Collection<? extends PhpNamedElement> tryFunctionCall( CustomMethodCallSignature customMethodCallSignature,
                                                                   String signature,
                                                                   PhpIndex phpIndex,
                                                                   Project project ) {
        SignatureMatcher signatureMatcher = new SignatureMatcher();
        if ( signatureMatcher.verifySignatureIsClassConstantFunctionCall( signature ) ) {
            return phpIndex.getAnyByFQN(
                    classConstantAnalyzer.getClassNameFromConstantLookup( signature, project )
            );
        }

        return tryToDeferToDefaultType( customMethodCallSignature, signature, phpIndex );
    }


    private Collection<? extends PhpNamedElement> tryToDeferToDefaultType( CustomMethodCallSignature customMethodCallSignature,
                                                                           String signature,
                                                                           PhpIndex phpIndex ) {
        if ( signature.indexOf( "#" ) != 0 ) {
            if ( signature.indexOf( "\\" ) != 0 ) {
                signature = "\\" + signature;
            }
            return phpIndex.getAnyByFQN( signature );
        }

        signature = cleanConstant( signature );

        try {
            return phpIndex.getBySignature( signature, null, 0 );
        } catch ( RuntimeException e ) {
            String signatureMessage = "function call";
            if ( customMethodCallSignature != null ) {
                signatureMessage = customMethodCallSignature.toString();
            }

            logger.warn(
                    "CustomSignatureProcessor.tryToDeferToDefaultType cannot decode " + signature + "\n" +
                            signatureMessage
            );

            return Collections.emptySet();
        }
    }


    /**
     * Indicates something is wrong from one of the tests in the test environment but no inspections are raised.
     * Just a warning in idea. Will need further looking into.
     *
     * @param signature
     * @return
     */
    private String cleanConstant( String signature ) {
        if ( signature.indexOf( "#K#C" ) == 0 && !signature.contains( "|?" ) ) {
            signature = signature + ".|?";
        }
        return signature;
    }
}

File: src/main/kotlin/com/ptby/dynamicreturntypeplugin/signatureconversion/CustomSignatureProcessor.kt
package com.ptby.dynamicreturntypeplugin.signatureconversion

import com.intellij.openapi.diagnostic.Logger.getInstance
import com.intellij.openapi.project.Project
import com.jetbrains.php.PhpIndex
import com.jetbrains.php.lang.psi.elements.PhpNamedElement
import com.ptby.dynamicreturntypeplugin.index.ClassConstantWalker
import com.ptby.dynamicreturntypeplugin.index.FieldReferenceAnalyzer
import com.ptby.dynamicreturntypeplugin.index.ReturnInitialisedSignatureConverter
import com.ptby.dynamicreturntypeplugin.index.VariableAnalyser
import com.ptby.dynamicreturntypeplugin.signature_extension.startsWithClassConstantPrefix
import com.ptby.dynamicreturntypeplugin.signature_processingv2.ListReturnPackaging

class CustomSignatureProcessor(private val returnInitialisedSignatureConverter: ReturnInitialisedSignatureConverter,
                                      private val classConstantWalker: ClassConstantWalker,
                                      private val fieldReferenceAnalyzer: FieldReferenceAnalyzer,
                                      private val variableAnalyser: VariableAnalyser) : ListReturnPackaging {
    private val logger = getInstance("DynamicReturnTypePlugin")


    fun processSignature(phpIndex: PhpIndex,
                         customMethodCallSignature: CustomMethodCallSignature,
                         project: Project,
                         signature: String): Collection<PhpNamedElement>? {

        var processedCustomMethodCallSignature = customMethodCallSignature
        val signatureMatcher = SignatureMatcher()
        if (signatureRequiresConversion(processedCustomMethodCallSignature, signatureMatcher)) {
            processedCustomMethodCallSignature = returnInitialisedSignatureConverter.convertSignatureToClassSignature(
                    processedCustomMethodCallSignature, project
            )
        }

        if (signatureMatcher.verifySignatureIsClassConstantFunctionCall(processedCustomMethodCallSignature)) {
            return phpIndex.getAnyByFQN(classConstantWalker.walkThroughConstants(
                    project,
                    processedCustomMethodCallSignature.rawStringSignature
            )
            )
        } else if (signatureMatcher.verifySignatureIsFieldCall(processedCustomMethodCallSignature)) {
            return fieldReferenceAnalyzer.getClassNameFromFieldLookup(processedCustomMethodCallSignature, project)
        } else if (signatureMatcher.verifySignatureIsMethodCall(processedCustomMethodCallSignature)) {
            return variableAnalyser.getClassNameFromVariableLookup(processedCustomMethodCallSignature, project)
        }

        return tryToDeferToDefaultType(processedCustomMethodCallSignature, signature, phpIndex)
    }

    private fun signatureRequiresConversion(processedCustomMethodCallSignature: CustomMethodCallSignature,
                                            signatureMatcher: SignatureMatcher): Boolean {
        return signatureMatcher.verifySignatureIsDeferredGlobalFunctionCall(processedCustomMethodCallSignature) ||
                signatureMatcher.verifySignatureIsFromReturnInitialiasedLocalObject(processedCustomMethodCallSignature)
    }


    fun tryFunctionCall(parameter: String,
                        phpIndex: PhpIndex,
                        project: Project): Collection<PhpNamedElement> {
        val signatureMatcher = SignatureMatcher()
        if (signatureMatcher.verifySignatureIsClassConstantFunctionCall(parameter)) {
            return phpIndex.getAnyByFQN(
                    classConstantWalker.walkThroughConstants( project, parameter)
            )
        }

        if ( requiresListPackaging(parameter)) {
            return packageList(parameter, project)
        }

        return tryToDeferToDefaultType(null, parameter, phpIndex)
    }


    private fun tryToDeferToDefaultType(customMethodCallSignature: CustomMethodCallSignature?,
                                        signature: String,
                                        phpIndex: PhpIndex): Collection<PhpNamedElement> {
        var cleanedSignature = signature.replace("\\\\", "\\")
        if (cleanedSignature.indexOf("#") != 0) {
            if (cleanedSignature.indexOf("\\") != 0) {
                cleanedSignature = "\\" + cleanedSignature
            }
            return phpIndex.getAnyByFQN(cleanedSignature)
        }

        cleanedSignature = cleanConstant(cleanedSignature)

        try {
            return phpIndex.getBySignature(cleanedSignature, null, 0)
        } catch (e: RuntimeException) {
            var signatureMessage = "function call"
            if (customMethodCallSignature != null) {
                signatureMessage = customMethodCallSignature.toString()
            }

            logger.warn(
                    "CustomSignatureProcessor.tryToDeferToDefaultType cannot decode " +
                            cleanedSignature + "\n"
                            + signatureMessage
            )

            return setOf()
        }
    }


    /**
     * Indicates something is wrong from one of the tests in the test environment but no inspections are raised.
     * Just a warning in idea. Will need further looking into.
     *
     * @param signature
     * @return
     */
    private fun cleanConstant(signature: String): String {
        var cleanedSignature = signature
        if (cleanedSignature.startsWithClassConstantPrefix() && !cleanedSignature.contains("|?")) {
            cleanedSignature += ".|?"
        }
        return cleanedSignature
    }
}
